### PR TITLE
Migrate from core native to Nokee plugins

### PR DESCRIPTION
**Pull request by lacasseio**
Fixes https://github.com/gradle/native-platform/issues/297.

See the attached issues for general work items left to do. Feel free to provide any kind of feedback on this PR either by starting a new conversation or contributing to a current conversation.

### CODE REVIEW

**Comment by lacasseio** [comment]
@big-guy At the moment, the branch is using a local version so the CI will fail. I will cut a release for you once our CI stabilizes.

File: buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import org.gradle.nativeplatform.toolchain.Clang;
+import org.gradle.nativeplatform.toolchain.Gcc;
+import org.gradle.nativeplatform.toolchain.NativeToolChainRegistry;
+import org.gradle.nativeplatform.toolchain.VisualCpp;
+
+import java.io.File;
+import java.util.Set;
+
+@SuppressWarnings("UnstableApiUsage")
+public abstract class JniNokeePlugin implements Plugin<Project> {
```
Comment (lacasseio) [inline buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java:34]: ❓ Is there a reason the JNI tests are run as part of another `Test` task? Nokee wires everything for the main `Test` task. There wasn't any consideration to "remove" the wiring of the JNI into the main test task. We can create a new test task but all the wiring will have to be done manually. Nokee could provide a `jni-unit-test` plugin that allows users to declare any number of JNI test tasks but that is a bit outside of the current scope. The project could remove the extra test task in favour of the main test task however there may be a reason for the task to existing in the first place.

**Comment by lacasseio** [comment]
@big-guy The dependency verification is quite unhelpful. Locally, it fails with an `Invalid UTF-8` error which is not clear as to what's the problem. I removed the verification files locally until we can work around this issue.

File: buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                project.getGroup().toString().replace('.', '/'),
+                "platform",
+                toVariantName(variant.getTargetMachine())));
+        });
+    }
+
+    private void registerWindowsDistributionDimension(JavaNativeInterfaceLibrary library) {
+        SetProperty<WindowsDistribution> newDimension = library.getDimensions().newAxis(WindowsDistribution.class, builder -> builder.onlyOn(library.getMachines().getWindows().getOperatingSystemFamily()));
+        newDimension.convention(ImmutableSet.copyOf(WindowsDistribution.values()));
+        ((ExtensionAware) library).getExtensions().add("targetWindowsDistributions", newDimension);
```
Comment (lacasseio) [inline buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java:121]: 🔍 See conversation in the `#ext-nokee` Slack channel around `ExtensionAware` guideline.

**Comment by lacasseio** [comment]
@big-guy I released an integration version. We still need to figure out the dependency verification error for CI to be happy.

File: buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java (hunk 1 above)
Comment (lptr) [reply buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java:36]: We have some special tests where `-Xcheck:jni` output is treated as a reason for failure. These are the tests marked with `JniChecksEnabled`. For other tests that were introduced earlier, we did not enforce these checks yet, and they do fail.

One option to remove this distinction would be to make sure none of the current code emits any JNI warnings.

File: buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        VariantsExtension variants = project.getExtensions().getByType(VariantsExtension.class);
+        variants.getVariantNames().set(library.getVariants().flatMap(it -> {
+            // Only depend on variants which can be built on the current machine
+            boolean onlyLocalVariants = project.getProviders().gradleProperty("onlyLocalVariants").forUseAtConfigurationTime().isPresent();
+            if (onlyLocalVariants && !it.getSharedLibrary().isBuildable()) {
+                return ImmutableList.of();
+            } else {
+                return ImmutableList.of(toVariantName(it.getTargetMachine()));
+            }
+        }));
```
Comment (big-guy) [inline buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java:66]: What does it mean to set the "variant names" here? This disables certain variants from running?

File: native-platform/build.gradle [hunk 4]
```diff
@@ -20,41 +23,15 @@ javadoc {
...
-                }
-            }
+library {
+    baseName = 'native-platform'
+    sources.configureEach(CppSourceSet) {
+        source.from('src/shared/cpp', 'src/main/cpp')
+        headers.from('src/shared/headers', 'build/generated/jni-headers/main')
+    }
+    variants.configureEach {
+        if (NcursesVersion.values().any { buildVariant.hasAxisOf(it) }) {
```
[the remark this answers is not in the document]
Comment (lacasseio) [reply native-platform/build.gradle:33]: It's a little bit out of date but it should give you a good enough idea: https://docs.nokee.dev/dsl/dev.nokee.platform.jni.JavaNativeInterfaceLibrary.html

**Pull request merged**